contracts/math/uints.ral

```diff
@@ -90,5 +90,74 @@ Abstract Contract U512Impl () {
                 lower: deltaL
             }
         }
+
+    pub fn bigAdd512(a: U512, b: U512) -> U512 {
+        let (r0, overflow) = overflowingAdd(a.lower, b.lower)
+        let r1 = wrappingAdd(wrappingAdd(a.higher, b.higher), overflow)
+        return U512 { higher: r1, lower: r0 }
+    }
+
+    pub fn bigMul(a: U256, b: U256, denominator: U256) -> U512 {
+        let aLower = low128(a)
+        let aHigher = high128(a)
+        let bLower = low128(b)
+        let bHigher = high128(b)
+
+        let aLowerBLower = aLower * bLower
+        let aHigherBHigher = aHigher * bHigher
+        let aHigherBLower = aHigher * bLower
+        let aLowerBHigher = aLower * bHigher
+
+        let mut result = bigAdd512(U512 { higher: aHigherBHigher, lower: aLowerBLower }, U512 { higher: high128(aHigherBLower), lower: low128(aHigherBLower) << 128 })
+        result = bigAdd512(result, U512 { higher: high128(aLowerBHigher), lower: low128(aLowerBHigher) << 128 })
+
+        result = bigDiv(result, denominator, 1)
+
+        return result
+    }
+
+    pub fn bigMulUp(a: U256, b: U256, denominator: U256) -> U512 {
```

Comment:
substantial portion of the code is shared between `bigMul` and `bigMulUp`. As is often the case, numerous points of definition can lead to unexpected inconsistencies if we only replace one part when making changes.

It's worth abstracting this common logic into a separate function